Migrate evohome, cache access tokens

zxdavb opened the pull request:
## Breaking Change:

Those migrating from **honeywell** to **evohome** will lose higher-precision temperatures (was 0.1C, now 0.5 for climate, and 1.0 for water_heater).

## Description:
This is a large PR , which requires/includes:
 - Cache access tokens to prevent exceeding (recently lowered) API limits during restarts of HA (was required for dev/testing for `climate-1.0`)
 - ~Adds a `water_heater` component to the integration (required as this support has been removed in #24257)~
 - Bump client library (fixes bug preventing use of refresh tokens as above)
 - Fix a bug where zone operating mode was mis-reported
 - Minor refactoring (see below)

**Note**: Without `water_heater` support, **evohome** is not suitable for those switching from **honeywell** (see #23469).

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** N/A

## Example entry for `configuration.yaml` (if applicable): Unchanged

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

zxdavb committed:
add water_heater, add storage - initial commit

delint

add missing code

desiderata

update honeywell client library & CODEOWNER

add auth_tokens code, refactor & delint

refactor for broker

delint

zxdavb committed:
add water_heater, add storage - initial commit

zxdavb committed:
Add Broker, Water Heater & Refactor

add missing code

desiderata

balloob commented on requirements_test_all.txt:
```diff
@@ -108,10 +108,12 @@ enocean==0.50
 evohomeclient==0.3.2
 
+# homeassistant.components.evohome
+evohomeclient==0.3.3
```
This won't work without also upgrading the version in the honeywell integration.

zxdavb replied:
woops - so busy with the detail, forgot the big things: thanks.  Added myself as codeowner too.

zxdavb committed:
update honeywell client library & CODEOWNER

add auth_tokens code, refactor & delint

refactor for broker

zxdavb commented:
@balloob this is WIP - has a lot of debug code, etc., that will be removed before I remove the WIP.

Are you OK with this going through before I submit another PR to migrate **evohome**?  @pvizeli wanted me to do both (and moved the earlier version of the PR across to `climate-1.0`), but it just wasn't working out for the reasons outlined above in the **Description**.

balloob commented:
Ha, it was more like a drive by comment. I didn't do a full review. Let me know when it's ready and I'll review it 👍 I think that having stuff moved to water heater is a good move, as climate-1 is going to break that stuff most likely.

zxdavb committed:
bugfix - loc_idx may not be 0

more refactor - ensure pure async

more refactoring

appears all r/o attributes are working

tweak precsion, DHW & delint

remove unused code

remove unused code 2

remove unused code, refactor _save_auth_tokens()

zxdavb committed:
support RoundThermostat

bugfix opmode, switch to util.dt, add until=1h

revert breaking change

zxdavb committed:
store at_expires as naive UTC

remove debug code

delint

tidy up exception handling

delint

add water_heater, add storage - initial commit

delint

add missing code

desiderata

update honeywell client library & CODEOWNER

add auth_tokens code, refactor & delint

refactor for broker

add water_heater, add storage - initial commit

delint

add missing code

desiderata

update honeywell client library & CODEOWNER

add auth_tokens code, refactor & delint

refactor for broker

delint

bugfix - loc_idx may not be 0

more refactor - ensure pure async

more refactoring

appears all r/o attributes are working

tweak precsion, DHW & delint

remove unused code

remove unused code 2

remove unused code, refactor _save_auth_tokens()

support RoundThermostat

bugfix opmode, switch to util.dt, add until=1h

revert breaking change

store at_expires as naive UTC

remove debug code

delint

tidy up exception handling

delint

zxdavb committed:
update CODEOWNERS

zxdavb commented:
Sorry for spamming reviewer - forgot to `git push --force`!

zxdavb commented on homeassistant/components/evohome/__init__.py:
```diff
@@ -44,204 +45,243 @@
+            at_expires_str = app_storage.get(CONF_ACCESS_TOKEN_EXPIRES)
+            if at_expires_str:
+                at_expires_dt = as_utc(parse_datetime(at_expires_str))
+                at_expires_dt = at_expires_dt.astimezone(tzlocal())
```
I tried to do this without using `dateutil.tztzlocal()` and instead using `util.dt.get_time_zone()`, but couldn't get it to work.

I am calling a web API and want to convert a local time (`access_token_expires`) into UTC (to be stored/restored, and) used for `async_track_point_in_utc_time()`

Only one other component uses `get_time_zone`, but gets the string from **configuration.yaml**, which is not satisfactory:
```python
./components/worldclock/sensor.py:30:    
time_zone = dt_util.get_time_zone(config.get(CONF_TIME_ZONE))
```
... as I need the offset used by the evohome API, and not the TZ as specified by the user.

balloob replied:
Is there any reason why `access_token_expires` is in local?

zxdavb replied:
Because local is the TZ used by the client API and the dt is naive.  The client API uses `access_token_expires` to know when to re-authorize, and the integration uses it to know when it is time to store the new tokens.

zxdavb committed:
fix regression

zxdavb commented:
> Let me know when it's ready and I'll review it 👍

@balloob ready for review, if you've time.

balloob commented on homeassistant/components/evohome/__init__.py:
```diff
@@ -44,204 +45,240 @@
-                self._timers['statusUpdated'] = datetime.now() + \
-                    self._params[CONF_SCAN_INTERVAL] * 3
+    async def _load_auth_tokens(self) -> Tuple[str, str, datetime]:
+        store = self.hass.helpers.storage.Store(STORAGE_VERSION, STORAGE_KEY)
```
This is literally why config entries exist, to manage auth tokens. We should probably just migrate directly to that .

zxdavb replied:
OMG - are you talking about that config flow thing? You couldn't sign-post me to an integration that is already doing it?

balloob replied:
There are 60 config flows to choose from 😉 

https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/generated/config_flows.py

An example is https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/met/config_flow.py#L166 and there is a ton of docs about it on https://developers.home-assistant.io/

zxdavb replied:
I just saw your post re: `climate-1.0`.

I am able to refactor 4 climate entities (**evohome**, **geniushub**, **honeywell**, **incomfort**), and have access to these systems fro testing. I have even submitted a PR for two of them.  

However, **evohome**, and **geniushub** wont be easy, and so I am waiting for feedback on the two I've already done (**honeywell**, **incomfort**), which haven't been merged, before I start work on them.

There is the added complicated of deprecating EU out of **honeywell**, and that obligates `water_heater` support in **evohome** (one of the two purposes of this PR)...

Additionally, the API was recently changed by the vendor, and rate limiting is on initial OAuth rather than ongoing polling - this means development/testing is virtually impossible unless access tokens can be cached/restored (the other purpose of this PR).

So...

Can I get this PR accepted as is, using storage, I'll then have time to work on `climate-1.0`, and after that, I'll go back and migrate to config entries?

Or is it all too late?

Otherwise, can I put all this code into `climate-1.0`? - will mean a large PR

balloob replied:
We can migrate later from storage.

balloob replied:
I'm also fine to take this as a big PR into climate 1.0.

zxdavb replied:
Great! Makes my life so much easier! I'll see if I can have this ready for review in <24h.

zxdavb commented:
I am going to rebase this against `climate-1.0`.

zxdavb committed:
fix requirements

zxdavb committed:
migrate to climate-1.0

zxdavb committed:
tweaking

zxdavb committed:
de-lint

zxdavb committed:
TCS working? & delint

zxdavb committed:
tweaking

zxdavb committed:
TCS code finalised

zxdavb committed:
remove available() logic

zxdavb committed:
refactor _switchpoints()

zxdavb committed:
tidy up switchpoint code

zxdavb committed:
tweak

zxdavb committed:
teaking device_state_attributes

balloob commented on homeassistant/components/climate/const.py:
```diff
@@ -45,6 +45,9 @@
 PRESET_COMFORT = 'comfort'
 
+# Device is in a user-configurable mode
+PRESET_CUSTOM = 'custom'
```
Let's not add this to the official constants. The official constants are there to standardize some things to help with translations/icons/Google/Alexa. Custom does not fit that bill and you can keep it inside evohome. There is no limitation on what values you return from the `preset_modes` property.

zxdavb replied:
OK, will do.  

I previously added `PRESET_ACTIVITY` for **geniushub** - it has movement sensors - is that OK?

zxdavb replied:
OK, thanks.

pvizeli replied:
yes, because that can be used by others

balloob commented on homeassistant/components/evohome/__init__.py:
```diff
@@ -44,229 +44,347 @@
+        hass,
+        broker.update,
+#       hass_config[DOMAIN][CONF_SCAN_INTERVAL]
+        timedelta(seconds=10)  # TODO: delete me
```
TODO

zxdavb replied:
Yeah, sorry - I am still in dev mode hence the **WIP:** in the title.

I though if I had **WIP:** in the title, then people would know the PR wasn't ready for review? But I had one merged!!

balloob commented on homeassistant/components/evohome/climate.py:
```diff
@@ -1,457 +1,334 @@
-            self.target_temperature == self.min_temp and \
-            self._status['setpointStatus']['setpointMode'] == EVO_PERMOVER
-        return not is_off
+    def preset_mode(self) -> Optional[str]:  # TODO: check
```
TODO

balloob commented on homeassistant/components/evohome/climate.py:
```diff
@@ -1,457 +1,334 @@
-        return not is_off
+    def preset_mode(self) -> Optional[str]:  # TODO: check
+        """Return the current preset mode, e.g., home, away, temp."""
+        return None
```
🐞

zxdavb replied:
WIP

balloob commented on homeassistant/const.py:
```diff
@@ -122,6 +123,7 @@
 CONF_QUOTE = 'quote'
 CONF_RADIUS = 'radius'
 CONF_RECIPIENT = 'recipient'
+CONF_REFRESH_TOKEN = 'refresh_token'
```
Please keep these constants local to the integration.

zxdavb replied:
I think that a `refresh_token`, being part of OAuth specification, would be OK for this. Same for `access_token_expires`?

I will take them out now, but will happily put them back if you change your mind.

zxdavb committed:
some refactoring

zxdavb committed:
move PRESET_CUSTOM back to evohome

zxdavb committed:
move CONF_ACCESS_TOKEN_EXPIRES CONF_REFRESH_TOKEN back to evohome

zxdavb committed:
refactor SP code and dt conversion

zxdavb committed:
delinted

zxdavb committed:
delinted

zxdavb committed:
remove water_heater

zxdavb committed:
fix regression

zxdavb commented:
@balloob This is ready for review - sorry it is so big.

MartinHjelmare commented on homeassistant/components/evohome/__init__.py:
```diff
@@ -44,229 +45,314 @@
-    if _LOGGER.isEnabledFor(logging.DEBUG):
-        tmp_loc = dict(evo_data['config'])
-        tmp_loc['locationInfo']['postcode'] = 'REDACTED'
+    load_platform(hass, 'climate', DOMAIN, {}, hass_config)
```
We're in async context so we have to use the home assistant async api.

`async_load_platform` should be sheduled as a task to avoid deadlock.

zxdavb replied:
OK, can submit a PR to address this this evening.

zxdavb replied:
```python
    hass.async_create_task(
         async_load_platform(hass, 'climate', DOMAIN, {}, hass_config))
```
